platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,231 @@
+- Create/load an initial state as a starting point. If no snapshot is available, it should create a new state. 
+- Managing references to the latest immutable state and the latest mutable state.
+- Restrict mutability to a single state. Specifically, only one state object should be mutable at a time. To achieve this, we need to use a combination of `StateLifecycleManager.copy` method and `NonCopyableState` class.
+- Load snapshots from the disk. A state loaded from a snapshot must be immutable.
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+  /**
+   * Get the latest immutable state.
+   */
+  State getLatestImmutableState();
+
+  /**
+   * Get the mutable state.
+   */
+  State getMutableState();
+
+  /**
+   * Load an immutable copy of a state from the snapshot at the specified path. 
+   * @param sourcePath the path to the snapshot
+   * @return the immutable state loaded from the snapshot
+   */
+  State loadSnapshot(Path sourcePath);
+
+  /**
+   * Calling this method creates a mutable copy of the state. The previous mutable state becomes immutable, 
+   * replacing the current latest immutable state, and is accessible via the {@link #getLatestImmutableState} method.
+   * @return a mutable copy of the state
+   */
+  State copy();
+}
+```
+
+The following diagram illustrates an example of the states in memory throughout their lifecycles:
+
+[![Multistate management](multi-states.svg)](multi-states.svg)
+
+### Restricting Copying
+
+To ensure that only `StateLifecycleManager` could manage mutability of the state object, a new implementation of the `State` interface must be introduced:
+
+```java
+public final class NonCopyableState implements State {
+    private State delegate; 
+    // All methods delegate to the delegate object except for the copy method.
+    // For the copy method, we retain the default implementation, which throws an `UnsupportedOperationException`.
+}
+```
+
+Client code requiring a mutable copy of the state must use the `copy` method provided by the `StateLifecycleManager` interface.
+
+### State Initialization
+
+The current state initialization logic is fragmented across multiple classes and methods, including:
+
+- `ServicesMain#main`
+- `Hedera#newMerkleStateRoot`
+- `StartupStateUtils.getInitialState`
+- `StartupStateUtils.loadStateFile`
+- `StartupStateUtils.loadLatestState`
+- `SignedStateFileReader.readStateFile`
+
+Additionally, this logic is tightly coupled with other functionalities, such as:
+
+- State signatures
+- Platform service registration
+- Invalid state file cleanup
+- Hash validation
+- State reservation
+
+Invalid state file cleanup and hash validation should be moved to the `StateLifecycleManager` implementation. The remaining responsibilities belong to the platform code.
```

Comment:
What kind of "invalid" checks? Will these checks be common to both CN and BN?

Reply:
I added a clarification for the "invalid" checks. Yes, they will be common for CN and BN.